File: go/core/website-analysis-apis2_sdk.go
package core

import (
	"fmt"

	vs "github.com/voxgig-sdk/website-analysis-apis2-sdk/go/utility/struct"
)

type WebsiteAnalysisApis2SDK struct {
	Mode     string
	options  map[string]any
	utility  *Utility
	Features []Feature
	rootctx  *Context
}

func NewWebsiteAnalysisApis2SDK(options map[string]any) *WebsiteAnalysisApis2SDK {
	sdk := &WebsiteAnalysisApis2SDK{
		Mode:     "live",
		Features: []Feature{},
	}

	sdk.utility = NewUtility()

	config := MakeConfig()

	sdk.rootctx = sdk.utility.MakeContext(map[string]any{
		"client":  sdk,
		"utility": sdk.utility,
		"config":  config,
		"options": options,
		"shared":  map[string]any{},
	}, nil)

	sdk.options = sdk.utility.MakeOptions(sdk.rootctx)

	if vs.GetPath([]any{"feature", "test", "active"}, sdk.options) == true {
		sdk.Mode = "test"
	}

	sdk.rootctx.Options = sdk.options

	// Add features in the resolved order (MakeOptions puts an explicit array
	// order first, else defaults to test-first). Ordering matters: the `test`
	// feature installs the base mock transport and the transport features
	// (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
	// must be added before them to sit at the base of the chain.
	featureOpts := ToMapAny(vs.GetProp(sdk.options, "feature"))
	if featureOpts != nil {
		if fo, ok := vs.GetPath([]any{"__derived__", "featureorder"}, sdk.options).([]any); ok {
			for _, n := range fo {
				fname, _ := n.(string)
				fopts := ToMapAny(featureOpts[fname])
				if fopts != nil {
					if active, ok := fopts["active"]; ok {
						if ab, ok := active.(bool); ok && ab {
							sdk.utility.FeatureAdd(sdk.rootctx, makeFeature(fname))
						}
					}
				}
			}
		}
	}

	// Add extension features.
	if extend := vs.GetProp(sdk.options, "extend"); extend != nil {
		if extList, ok := extend.([]any); ok {
			for _, f := range extList {
				if feat, ok := f.(Feature); ok {
					sdk.utility.FeatureAdd(sdk.rootctx, feat)
				}
			}
		}
	}

	// Initialize features.
	for _, f := range sdk.Features {
		sdk.utility.FeatureInit(sdk.rootctx, f)
	}

	sdk.utility.FeatureHook(sdk.rootctx, "PostConstruct")

	return sdk
}

func (sdk *WebsiteAnalysisApis2SDK) OptionsMap() map[string]any {
	out := vs.Clone(sdk.options)
	if om, ok := out.(map[string]any); ok {
		return om
	}
	return map[string]any{}
}

func (sdk *WebsiteAnalysisApis2SDK) GetUtility() *Utility {
	return CopyUtility(sdk.utility)
}

func (sdk *WebsiteAnalysisApis2SDK) GetRootCtx() *Context {
	return sdk.rootctx
}

func (sdk *WebsiteAnalysisApis2SDK) Prepare(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "prepare",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	options := sdk.options

	path, _ := vs.GetProp(fetchargs, "path").(string)
	method, _ := vs.GetProp(fetchargs, "method").(string)
	if method == "" {
		method = "GET"
	}

	params := ToMapAny(vs.GetProp(fetchargs, "params"))
	if params == nil {
		params = map[string]any{}
	}
	query := ToMapAny(vs.GetProp(fetchargs, "query"))
	if query == nil {
		query = map[string]any{}
	}

	headers := utility.PrepareHeaders(ctx)

	base, _ := vs.GetProp(options, "base").(string)
	prefix, _ := vs.GetProp(options, "prefix").(string)
	suffix, _ := vs.GetProp(options, "suffix").(string)

	ctx.Spec = NewSpec(map[string]any{
		"base":    base,
		"prefix":  prefix,
		"suffix":  suffix,
		"path":    path,
		"method":  method,
		"params":  params,
		"query":   query,
		"headers": headers,
		"body":    vs.GetProp(fetchargs, "body"),
		"step":    "start",
	})

	// Merge user-provided headers.
	if uh := vs.GetProp(fetchargs, "headers"); uh != nil {
		if uhm, ok := uh.(map[string]any); ok {
			for k, v := range uhm {
				ctx.Spec.Headers[k] = v
			}
		}
	}

	_, err := utility.PrepareAuth(ctx)
	if err != nil {
		return nil, err
	}

	return utility.MakeFetchDef(ctx)
}

func (sdk *WebsiteAnalysisApis2SDK) Direct(fetchargs map[string]any) (map[string]any, error) {
	utility := sdk.utility

	fetchdef, err := sdk.Prepare(fetchargs)
	if err != nil {
		return map[string]any{"ok": false, "err": err}, nil
	}

	if fetchargs == nil {
		fetchargs = map[string]any{}
	}

	var ctrl map[string]any
	if c := vs.GetProp(fetchargs, "ctrl"); c != nil {
		if cm, ok := c.(map[string]any); ok {
			ctrl = cm
		}
	}
	if ctrl == nil {
		ctrl = map[string]any{}
	}

	ctx := utility.MakeContext(map[string]any{
		"opname": "direct",
		"ctrl":   ctrl,
	}, sdk.rootctx)

	url, _ := fetchdef["url"].(string)
	fetched, fetchErr := utility.Fetcher(ctx, url, fetchdef)

	if fetchErr != nil {
		return map[string]any{"ok": false, "err": fetchErr}, nil
	}

	if fetched == nil {
		return map[string]any{
			"ok":  false,
			"err": ctx.MakeError("direct_no_response", "response: undefined"),
		}, nil
	}

	if fm, ok := fetched.(map[string]any); ok {
		status := ToInt(vs.GetProp(fm, "status"))
		headers := vs.GetProp(fm, "headers")

		// No-body responses (204, 304) and explicit zero content-length
		// must skip JSON parsing — calling json() on an empty body errors.
		var contentLength string
		if hm, ok := headers.(map[string]any); ok {
			if cl, ok := hm["content-length"]; ok {
				contentLength = fmt.Sprintf("%v", cl)
			}
		}
		noBody := status == 204 || status == 304 || contentLength == "0"

		var jsonData any
		if !noBody {
			if jf := vs.GetProp(fm, "json"); jf != nil {
				if f, ok := jf.(func() any); ok {
					// f() returns nil on parse error in our fetcher.
					jsonData = f()
				}
			}
		}

		return map[string]any{
			"ok":      status >= 200 && status < 300,
			"status":  status,
			"headers": headers,
			"data":    jsonData,
		}, nil
	}

	return map[string]any{"ok": false, "err": ctx.MakeError("direct_invalid", "invalid response type")}, nil
}


// DeadLinkChecker returns a DeadLinkChecker entity bound to this client.
// Idiomatic usage: client.DeadLinkChecker(nil).List(nil, nil) or
// client.DeadLinkChecker(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *WebsiteAnalysisApis2SDK) DeadLinkChecker(data map[string]any) WebsiteAnalysisApis2Entity {
	return NewDeadLinkCheckerEntityFunc(sdk, data)
}


// Performance returns a Performance entity bound to this client.
// Idiomatic usage: client.Performance(nil).List(nil, nil) or
// client.Performance(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *WebsiteAnalysisApis2SDK) Performance(data map[string]any) WebsiteAnalysisApis2Entity {
	return NewPerformanceEntityFunc(sdk, data)
}


// Screenshot returns a Screenshot entity bound to this client.
// Idiomatic usage: client.Screenshot(nil).List(nil, nil) or
// client.Screenshot(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *WebsiteAnalysisApis2SDK) Screenshot(data map[string]any) WebsiteAnalysisApis2Entity {
	return NewScreenshotEntityFunc(sdk, data)
}


// Seo returns a Seo entity bound to this client.
// Idiomatic usage: client.Seo(nil).List(nil, nil) or
// client.Seo(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *WebsiteAnalysisApis2SDK) Seo(data map[string]any) WebsiteAnalysisApis2Entity {
	return NewSeoEntityFunc(sdk, data)
}


// Ssl returns a Ssl entity bound to this client.
// Idiomatic usage: client.Ssl(nil).List(nil, nil) or
// client.Ssl(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *WebsiteAnalysisApis2SDK) Ssl(data map[string]any) WebsiteAnalysisApis2Entity {
	return NewSslEntityFunc(sdk, data)
}


// TechStackDetection returns a TechStackDetection entity bound to this client.
// Idiomatic usage: client.TechStackDetection(nil).List(nil, nil) or
// client.TechStackDetection(nil).Load(map[string]any{"id": ...}, nil).
func (sdk *WebsiteAnalysisApis2SDK) TechStackDetection(data map[string]any) WebsiteAnalysisApis2Entity {
	return NewTechStackDetectionEntityFunc(sdk, data)
}



func TestSDK(testopts map[string]any, sdkopts map[string]any) *WebsiteAnalysisApis2SDK {
	if sdkopts == nil {
		sdkopts = map[string]any{}
	}
	sdkopts = vs.Clone(sdkopts).(map[string]any)

	if testopts == nil {
		testopts = map[string]any{}
	}
	testopts = vs.Clone(testopts).(map[string]any)
	testopts["active"] = true

	vs.SetPath(sdkopts, []any{"feature", "test"}, testopts)

	sdk := NewWebsiteAnalysisApis2SDK(sdkopts)
	sdk.Mode = "test"

	return sdk
}

File: ts/src/WebsiteAnalysisApis2SDK.ts
// WebsiteAnalysisApis2 Ts SDK

import { DeadLinkCheckerEntity } from './entity/DeadLinkCheckerEntity'
import { PerformanceEntity } from './entity/PerformanceEntity'
import { ScreenshotEntity } from './entity/ScreenshotEntity'
import { SeoEntity } from './entity/SeoEntity'
import { SslEntity } from './entity/SslEntity'
import { TechStackDetectionEntity } from './entity/TechStackDetectionEntity'

export type * from './WebsiteAnalysisApis2Types'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { WebsiteAnalysisApis2EntityBase } from './WebsiteAnalysisApis2EntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class WebsiteAnalysisApis2SDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    // Add features in the resolved order (makeOptions puts an explicit
    // array order first, else defaults to test-first). Ordering matters:
    // the `test` feature installs the base mock transport and the transport
    // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is current,
    // so `test` must be added before them to sit at the base of the chain.
    const featureorder = getpath(this._options, '__derived__.featureorder') || []
    for (const fname of featureorder) {
      const fopts = this._options.feature[fname] || {}
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    }

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  // Entity access: `client.DeadLinkChecker().list()` / `client.DeadLinkChecker().load({ id })`.
  DeadLinkChecker(data?: any) {
    const self = this
    return new DeadLinkCheckerEntity(self,data)
  }


  // Entity access: `client.Performance().list()` / `client.Performance().load({ id })`.
  Performance(data?: any) {
    const self = this
    return new PerformanceEntity(self,data)
  }


  // Entity access: `client.Screenshot().list()` / `client.Screenshot().load({ id })`.
  Screenshot(data?: any) {
    const self = this
    return new ScreenshotEntity(self,data)
  }


  // Entity access: `client.Seo().list()` / `client.Seo().load({ id })`.
  Seo(data?: any) {
    const self = this
    return new SeoEntity(self,data)
  }


  // Entity access: `client.Ssl().list()` / `client.Ssl().load({ id })`.
  Ssl(data?: any) {
    const self = this
    return new SslEntity(self,data)
  }


  // Entity access: `client.TechStackDetection().list()` / `client.TechStackDetection().load({ id })`.
  TechStackDetection(data?: any) {
    const self = this
    return new TechStackDetectionEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new WebsiteAnalysisApis2SDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return WebsiteAnalysisApis2SDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'WebsiteAnalysisApis2' }
  }

  toString() {
    return 'WebsiteAnalysisApis2 ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = WebsiteAnalysisApis2SDK


export {
  stdutil,
  config,

  BaseFeature,
  WebsiteAnalysisApis2EntityBase,

  WebsiteAnalysisApis2SDK,
  SDK,
}


